### ISSUE TITLE
Access pattern information (line number

### ISSUE THREAD

**Issue by patrickt** [tree-sitter/tree-sitter#1066]

**Comment by maxbrunsfeld** [comment]
Currently, we [store the *byte offset*](https://github.com/tree-sitter/tree-sitter/blob/ca6e8d934155b479161573ac5a63f5e56e1a1374/lib/src/query.c#L122) of each pattern in the source file, and [expose that](https://github.com/tree-sitter/tree-sitter/blob/ca6e8d934155b479161573ac5a63f5e56e1a1374/lib/src/query.c#L2197-L2202) in the C API. In at least one place, the the rust bindings will use a byte offset to [re-derive the line number](https://github.com/tree-sitter/tree-sitter/blob/ca6e8d934155b479161573ac5a63f5e56e1a1374/lib/binding_rust/lib.rs#L1191-L1203). I'd be open to just storing the line numbers as well, so that they could be included in the LOG messages.

**Comment by dcreager** [comment]
Another interesting possibility would be to give each query pattern an optional textual label.  We could add some syntax to let you specify the name explicitly, for instance the `<function_call>` in the following:

```
<function_call>
(call_expression
  function: [
    (identifier) @name
    (parenthesized_expression (identifier) @name)
    (selector_expression field: (field_identifier) @name)
    (parenthesized_expression (selector_expression field: (field_identifier) @name))
  ]) @reference.call
```

And there could be a useful fallback, where if a pattern has only one capture, then we automatically use the capture name as the name of the pattern, assuming that name has already been used earlier in the file.

**Comment by patrickt** [comment]
@dcreager Yeah, I was thinking that the most-highly-advanced version of this looks like being able to add metadata (in the Clojure sense) to a given pattern.

**Comment by dcreager** [comment]
I think that might go too far — a simple textual label seems like a better sweet spot between "integer index" and "arbitrary bag of properties".  You already can use the integer index into your own bag of properties if you choose to create a more complex syntax that builds on the query language.  This would be used in the lowest level (read, implemented in C) query execution code, and so needs to be as simple as possible.

**Comment by maxbrunsfeld** [comment]
> give each query pattern an optional textual label

Yeah, That's interesting. Would this just be for the debug logging via that `LOG` macro, or are you thinking about other possible use-cases as well? If it's just for the debugging, I might rather start with something that doesn't affect the actual query input syntax, like logging out the line number, and possibly the root node name of the pattern.

I'm definitely open to enhancing the query syntax too though, if you think it makes sense. If we go down that road, I do kinda like @patrickt 's point about "metadata", in that, if we add new syntax, it'd be nice if there were some syntactic "precedent" for it in other lisps, or at least existing lisp parsers/syntax-highlighters wouldn't be thrown off by it.

**Comment by amaanq** [comment]
I think adding the line/number info would unnecessarily double the size of step offsets for little gain, and doesn't really have a use case outside of debugging. For metadata, we can already do that with predicates and directives (namely, `#set!`).
